sensor/common/sensor/central_communication_test.go

```diff
@@ -198,28 +197,65 @@ func (c *centralCommunicationSuite) Test_ClientReconciliation() {
 				responsesC <- message.New(msg)
 			}
 
-			if tc.eventuallyMatchesState != nil {
-				c.Assert().Eventually(func() bool {
-					mu.Lock()
-					defer mu.Unlock()
-					return tc.eventuallyMatchesState(copySentMessages(sentMessages))
-				}, 300*time.Millisecond, 10*time.Millisecond)
-			}
-
-			if tc.neverMatchesState != nil {
-				c.Assert().Never(func() bool {
-					mu.Lock()
-					defer mu.Unlock()
-					return tc.neverMatchesState(copySentMessages(sentMessages))
-				}, 300*time.Millisecond, 10*time.Millisecond)
+			select {
```

Comment:
Why not use `c.controller.Finish()` or something like:
```golang
c.Eventually(func() bool {
	return c.controller.Satisfied()
}, 100*time.Millisecond, 10*time.Millisecond)
```

If all test expectations are in gomock than this should be enough to check if the assertion pased.

Reply:
I was having weird data races using `Eventually` and the table tests. Let me give a try to your suggestion and see if the data races come back or not.

Reply:
I think I found the problem. I wrote this simple test to illustrate it: 

```golang
func (c *centralCommunicationSuite) Test_DataRace() {
	_, closeFn := c.createCentralCommunication(true)
	defer closeFn()
	start := func() {
		_ = c.mockService.client.Send(nil)
		for {
			time.Sleep(10 * time.Millisecond)
			_, _ = c.mockService.client.Recv()
		}
	}
	go start()
	c.mockService.client.EXPECT().Send(gomock.Any()).Times(1)
	c.mockService.client.EXPECT().Recv().AnyTimes()
	time.Sleep(100 * time.Millisecond)
	c.controller.Finish()
}
```

If my conclusion is correct, the problem is that internally the goroutine that started in `Start` has `for` loops with `Recv` and `Context`. The test expects calls to these functions `Anytimes` [here](https://github.com/stackrox/stackrox/blob/a514ea04d38313b76780bea7d28dbce0c20889bf/sensor/common/sensor/central_communication_test.go#L278-L285). These functions write/read the mock `controller`, thus triggering the race if we use `Finish` or `Eventually`.

Reply:
If you are ok with it, I'll go ahead and merge this as it is.

Reply:
That's because all the `EXPECT` calls should be made in the beginning of the test. The snippet you commented above fails because you can't guarantee that the `EXPECT()` will be called before the actual function call under `go start()`. 

This test passes without any data race:
```go
func (c *centralCommunicationSuite) Test_DataRace() {
	_, closeFn := c.createCentralCommunication(true)
	c.mockService.client.EXPECT().Send(gomock.Any()).Times(1)
	c.mockService.client.EXPECT().Recv().AnyTimes()

	defer closeFn()
	start := func() {
		_ = c.mockService.client.Send(nil)
		for {
			time.Sleep(10 * time.Millisecond)
			_, _ = c.mockService.client.Recv()
		}
	}
	go start()

	time.Sleep(100 * time.Millisecond)
	c.controller.Finish()
}
```

Reply:
Mmm It is true that was a problem on my side with the snippet but the actual test is in the correct order.

If I run `go test -race -count=100 github.com/stackrox/rox/sensor/common/sensor -testify.m ^Test_DataRace` with your snippet I get the same data race:

```
=== RUN   Test_CentralCommunicationSuite
=== RUN   Test_CentralCommunicationSuite/Test_DataRace
--- PASS: Test_CentralCommunicationSuite (0.11s)
    --- PASS: Test_CentralCommunicationSuite/Test_DataRace (0.10s)
=== RUN   Test_CentralCommunicationSuite
=== RUN   Test_CentralCommunicationSuite/Test_DataRace
==================
WARNING: DATA RACE
Read at 0x00c00133dda8 by goroutine 58:
  github.com/stackrox/rox/sensor/common/sensor.(*centralCommunicationSuite).Test_DataRace.func1()
      /Users/lvalerom/go/src/github.com/stackrox/stackrox/sensor/common/sensor/central_communication_test.go:149 +0x7c

Previous write at 0x00c00133dda8 by goroutine 57:
  github.com/stackrox/rox/sensor/common/sensor.(*centralCommunicationSuite).createCentralCommunication.func1()
      /Users/lvalerom/go/src/github.com/stackrox/stackrox/sensor/common/sensor/central_communication_test.go:357 +0x4b
  runtime.deferreturn()
      /usr/local/go/src/runtime/panic.go:476 +0x32
  runtime.call16()
      /usr/local/go/src/runtime/asm_amd64.s:728 +0x48
  reflect.Value.Call()
      /usr/local/go/src/reflect/value.go:370 +0xc7
  github.com/stretchr/testify/suite.Run.func1()
      /Users/lvalerom/go/pkg/mod/github.com/stretchr/testify@v1.8.4/suite/suite.go:197 +0x70d
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1629 +0x47

Goroutine 58 (running) created at:
  github.com/stackrox/rox/sensor/common/sensor.(*centralCommunicationSuite).Test_DataRace()
      /Users/lvalerom/go/src/github.com/stackrox/stackrox/sensor/common/sensor/central_communication_test.go:152 +0x271
  runtime.call16()
      /usr/local/go/src/runtime/asm_amd64.s:728 +0x48
  reflect.Value.Call()
      /usr/local/go/src/reflect/value.go:370 +0xc7
  github.com/stretchr/testify/suite.Run.func1()
      /Users/lvalerom/go/pkg/mod/github.com/stretchr/testify@v1.8.4/suite/suite.go:197 +0x70d
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1629 +0x47

Goroutine 57 (finished) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1629 +0x805
  github.com/stretchr/testify/suite.runTests()
      /Users/lvalerom/go/pkg/mod/github.com/stretchr/testify@v1.8.4/suite/suite.go:242 +0x19c
  github.com/stretchr/testify/suite.Run()
      /Users/lvalerom/go/pkg/mod/github.com/stretchr/testify@v1.8.4/suite/suite.go:215 +0x9b2
  github.com/stackrox/rox/sensor/common/sensor.Test_CentralCommunicationSuite()
      /Users/lvalerom/go/src/github.com/stackrox/stackrox/sensor/common/sensor/central_communication_test.go:65 +0x44
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1629 +0x47
==================
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x103893b0b]

goroutine 90 [running]:
github.com/stackrox/rox/sensor/common/sensor.(*centralCommunicationSuite).Test_DataRace.func1()
	/Users/lvalerom/go/src/github.com/stackrox/stackrox/sensor/common/sensor/central_communication_test.go:149 +0x8b
created by github.com/stackrox/rox/sensor/common/sensor.(*centralCommunicationSuite).Test_DataRace
	/Users/lvalerom/go/src/github.com/stackrox/stackrox/sensor/common/sensor/central_communication_test.go:152 +0x272
FAIL	github.com/stackrox/rox/sensor/common/sensor	1.789s
FAIL
```

I think the problem is not stopping the goroutine. This passes:

```golang
func (c *centralCommunicationSuite) Test_DataRace() {
	for i := 0; i < 4; i++ { // emulate loop in table test
		c.Run(fmt.Sprintf("test %d", i), func() {
			_, closeFn := c.createCentralCommunication(true)
			defer closeFn()
			c.mockService.client.EXPECT().Send(gomock.Any()).Times(1)
			c.mockService.client.EXPECT().Recv().AnyTimes()

			stopSig := concurrency.NewSignal()
			sigDone := concurrency.NewSignal()
			start := func() {
				_ = c.mockService.client.Send(nil)
				for {
					select {
					case <-stopSig.Done():
						sigDone.Signal()
						return
					default:
						time.Sleep(10 * time.Millisecond)
						_, _ = c.mockService.client.Recv()
					}
				}
			}
			go start()

			time.Sleep(100 * time.Millisecond)
			c.controller.Finish()
			stopSig.Signal()
			sigDone.Wait()
		})
	}
}
```

I'll try stopping `CentralCommunication` in the test, hopefully that should fix the data race.